app/services/foreman_ansible/roles_importer.rb

```diff
@@ -1,24 +1,69 @@
 module ForemanAnsible
-  # Service to import roles from the filesystem.
-  # Should be extracted to a gem that both the proxy and Foreman can use.
+  # imports roles from smart proxy
   class RolesImporter
-    class << self
-      def import!
-        list_roles_in_fs.each do |role_name|
-          role = AnsibleRole.new(:name => role_name)
-          next if role.save
-          Rails.logger.debug("Failed to save role #{role_name}: "\
-                            "#{role.errors.full_messages.join(', ')}")
-        end
+    attr_reader :ansible_proxy
+
+    def initialize(proxy = nil)
+      @ansible_proxy = proxy
+    end
+
+    def import!
+      return import_roles remote_roles if ansible_proxy
+      import_roles local_roles
+    end
+
+    def import_roles(roles)
+      imported = roles.map do |role_name|
+        ::AnsibleRole.find_or_initialize_by(:name => role_name)
+      end
+      detect_changes imported
+    end
+
+    def finish_import(changes)
+      return unless  changes.present?
+      create_new_roles changes['new'] if changes['new']
+      delete_old_roles changes['obsolete'] if changes['obsolete']
```

Comment:
It'd be nice to show in this table if there are hosts/hostgroups associated with these hosts, maybe we can do it later